base/dates/io.jl

```diff
@@ -1,244 +1,248 @@
+@inline function tryparsenext(d::DatePart{'s'}, str, i, len)
+    ms, ii = tryparsenext_base10(str, i, len, min_width(d), max_width(d))
+    if !isnull(ms)
+        ms = Nullable{Int64}(get(ms) * 10 ^ (3 - (ii - i)))
     end
+    return ms, ii
 end
-slotparse(slot::Slot{Millisecond},x,locale) = !ismatch(r"[^0-9\s]",x) ? slot.parser(Base.parse(Float64,"."*x)*1000.0) : throw(SLOTERROR)
-slotparse(slot::Slot{DayOfWeekSlot},x,locale) = nothing
 
-function getslot(x,slot::DelimitedSlot,locale,cursor)
-    endind = first(search(x,slot.transition,nextind(x,cursor)))
-    if endind == 0 # we didn't find the next delimiter
-        s = x[cursor:end]
-        index = endof(x)+1
-    else
-        s = x[cursor:(endind-1)]
-        index = nextind(x,endind)
+
+### Format tokens
+
+for (c, fn) in zip("YmdHMS", [year, month, day, hour, minute, second])
+    @eval function format(io, d::DatePart{$c}, dt)
+        write(io, dec($fn(dt), d.width))
     end
-    return index, slotparse(slot,s,locale)
 end
-getslot(x,slot,locale,cursor) = (cursor+slot.width, slotparse(slot,x[cursor:(cursor+slot.width-1)], locale))
 
-function parse(x::AbstractString,df::DateFormat)
-    x = strip(x)
-    startswith(x, df.prefix) && (x = replace(x, df.prefix, "", 1))
-    isempty(x) && throw(ArgumentError("Cannot parse empty format string"))
-    if isa(df.slots[1], DelimitedSlot) && first(search(x,df.slots[1].transition)) == 0
-        throw(ArgumentError("Delimiter mismatch. Couldn't find first delimiter, \"$(df.slots[1].transition)\", in date string"))
-    end
-    periods = Period[]
-    extra = Any[]  # Supports custom slot types such as TimeZone
-    cursor = 1
-    for slot in df.slots
-        cursor, pe = getslot(x,slot,df.locale,cursor)
-        pe !== nothing && (isa(pe,Period) ? push!(periods,pe) : push!(extra,pe))
-        cursor > endof(x) && break
+for (tok, fn) in zip("uU", [monthabbr, monthname])
+    @eval function format(io, d::DatePart{$tok}, dt, locale)
+        write(io, $fn(month(dt), locale))
     end
-    sort!(periods,rev=true,lt=periodisless)
-    if isempty(extra)
-        return periods
-    else
-        return vcat(periods, extra)
+end
+
+for (tok, fn) in zip("eE", [dayabbr, dayname])
+    @eval function format(io, ::DatePart{$tok}, dt,  locale)
+        write(io, $fn(dayofweek(dt), locale))
     end
 end
 
-slotformat(slot,dt,locale) = lpad(string(value(slot.parser(dt))),slot.width,"0")
-function slotformat(slot::Slot{Year},dt,locale)
-    s = lpad(string(value(slot.parser(dt))),slot.width,"0")
-    if slot.letter == 'y'
-        return s[(end-slot.width+1):end]  # Truncate the year
-    else # == 'Y'
-        return s
+function format(io, d::DatePart{'y'}, dt)
+    y = year(dt)
+    n = d.width
+
+    # the last `n` digits of `y`
+    # will be 0 padded if `y` has less than `n` digits
+    str = dec(y,n)[end-(n-1):end]
+    write(io, str)
+end
+
+function format(io, d::DatePart{'s'}, dt)
+    write(io, string(millisecond(dt)/1000)[3:end])
```

Comment:
I think that would work; in Feather, I resorted to always writing a `view` of the underlying bytes, but a Substring should work too.